### PR TITLE
[Muon Slice] Orthogonalise allgather-EP expert weights on the full intermediate dim

### PR DESCRIPTION
**Pull request by xxyux**
### PR Category

Execute Infrastructure

### PR Types

Bug fixes

#### Problem

  With `moe_token_dispatcher_type="allgather"` and EP > 1, every rank holds all experts but only `moe_intermediate_size // EP` of each one's intermediate dim (`weight1: [E, H, 2I/EP]`, `weight2: [E, I/EP, H]`).

  `GroupedMLPExpert.muon_slice_specs` orthogonalised those local shards directly, so Newton-Schulz ran on a slab of each expert matrix instead of the matrix itself — the Muon update did not match a non-EP run.

  #### Fix

  - `muon_utils.ortho_ep_full_intermediate`: all-to-all into the traditional EP layout (`E // EP` experts, full intermediate), orthogonalise there, send the result back the same way.
  - `GroupedMLPExpert` routes `weight1` / `weight2` through it when its intermediate dim is EP-sharded; other layouts are unchanged.
  - `muon_ffn_split` still controls whether gate and up are orthogonalised independently or as one fused matrix.

  `SonicMoEExpert` inherits the fix.

### 是否引起精度变化

是

### CODE REVIEW

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 1]
```diff
@@ -284,16 +290,41 @@ def muon_slice_specs(self, muon_configs):
...
         from paddlefleet.transformer.muon_utils import (
+            ortho_ep_full_intermediate,
             ortho_gate_up,
             ortho_stacked,
         )
 
+        gate_up_fused = self.config.gated_linear_unit
+        ffn_split = muon_configs.get("muon_ffn_split", False)
+
+        if self.intermediate_ep_sharded:
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/moe/moe_expert.py:317]: P1 这里会让现有 `tests/single_card_tests/transformer/test_muon_slice_specs.py::TestGroupedMLPExpertSpecs` 直接失败：其中两个测试都用仅包含 `config` 的 `SimpleNamespace` 调用该方法，因此访问 `self.intermediate_ep_sharded` 会先抛出 `AttributeError`，原有断言无法执行。请同步补齐非 EP/EP fixture，或在这里兼容读取缺省的非 EP 状态，并确保现有用例继续通过。

File: src/paddlefleet/transformer/muon_utils.py [hunk 2]
```diff
@@ -66,6 +67,96 @@ def ortho_stacked(weight, ortho_fn):
...
+    if gate_up and weight.shape[shard_axis] % 2 != 0:
+        raise ValueError(
+            f"Fused gate/up axis must be even, got shape {weight.shape}."
+        )
+
+    # Splitting a contiguous tensor along axis 0 yields contiguous blocks, so
+    # the forward all-to-all needs no extra copy.
+    send = list(paddle.split(weight, ep_size, axis=0))
+    recv = [paddle.empty_like(send[0]) for _ in send]
+    dist.alltoall(recv, send, group=ep_group)
```
Comment (risemeup1111) [inline src/paddlefleet/transformer/muon_utils.py:119]: P1 这里的双向 `alltoall` 决定 expert、rank 以及 gate/up 分片的重排，顺序或 shape 出错会静默改变 Muon 更新；但本 PR 没有新增测试，现有 `test_sonic_moe_ep.py` 也未调用 Muon 或这个 helper。请补一个真实 EP=2 的多卡回归：为各 rank 构造可区分的中间维分片，并将 `weight1`/`weight2`（以及 `muon_ffn_split` 开/关）的返回结果与非 EP 的完整矩阵正交化参考逐 rank 对齐，覆盖正反两次通信。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
| 序号 | 位置 | 优先级 | 状态 |
|---|---|---|---|
| 1 | [moe_expert.py:309](https://github.com/PaddlePaddle/PaddleFleet/pull/1703#discussion_r3755008792) | P1 | ✅ |
| 2 | [muon_utils.py:119](https://github.com/PaddlePaddle/PaddleFleet/pull/1703#discussion_r3755009546) | P1 | ✅ |

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]

File: src/paddlefleet/transformer/moe/moe_expert.py [hunk 3]
```diff
@@ -197,6 +197,12 @@ def __init__(
         self.expert_parallel = (
             utils.get_pg_size(self.ep_group) > 1 if self.ep_group else False
         )
+        # Only the allgather-dispatcher EP path passes an explicit
+        # ``intermediate_size_per_partition`` (see MoELayer), i.e. this rank
+        # holds all experts but only I // EP of each one's intermediate dim.
+        self.intermediate_ep_sharded = (
+            intermediate_size_per_partition is not None and self.expert_parallel
+        )
```
Comment (adam-xiaoyao) [inline src/paddlefleet/transformer/moe/moe_expert.py]: 建议这里的判断条件换一下， 最好和def sharded_state_dict(...）里的is_intermediate_sharded判断统一

**Review by risemeup1111 — APPROVED** [review]

**Review by From00 — APPROVED** [review]
LGTM
